### DISCUSSION TITLE
ERROR request license token faild (3)

### DISCUSSION THREAD

**Discussion by sana888999** [dashingsoft/pyarmor#1203] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/67665857/235368616-02c5148f-3c3e-426c-86c5-eceb4352f8ca.png)
![image](https://user-images.githubusercontent.com/67665857/235368800-abe26480-310f-4213-8670-8267b7322347.png)

I just bought Lic last night been trying to register no luck. I keep getting error i tried everything..
I am tired of going through issues with no valid solution. Please advice!  

------------------------------------

8840 MainProcess 2023-04-30 10:52:50,681
Traceback (most recent call last):
  File "C:\Users\sana8\AppData\Local\Programs\Python\Python311\Lib\site-packages\pyarmor\cli\__main__.py", line 689, in main
    main_entry(sys.argv[1:])
  File "C:\Users\sana8\AppData\Local\Programs\Python\Python311\Lib\site-packages\pyarmor\cli\__main__.py", line 677, in main_entry
    args.func(ctx, args)
  File "C:\Users\sana8\AppData\Local\Programs\Python\Python311\Lib\site-packages\pyarmor\cli\__main__.py", line 280, in cmd_reg
    reg.register_regfile(regfile)
  File "C:\Users\sana8\AppData\Local\Programs\Python\Python311\Lib\site-packages\pyarmor\cli\register.py", line 163, in register_regfile
    self.update_token()
  File "C:\Users\sana8\AppData\Local\Programs\Python\Python311\Lib\site-packages\pyarmor\cli\register.py", line 111, in update_token
    Pytransform3._update_token(self.ctx)
  File "C:\Users\sana8\AppData\Local\Programs\Python\Python311\Lib\site-packages\pyarmor\cli\core\__init__.py", line 89, in _update_token
    m = Pytransform3.init(ctx)
        ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sana8\AppData\Local\Programs\Python\Python311\Lib\site-packages\pyarmor\cli\core\__init__.py", line 59, in init
    m.init_ctx(ctx)
RuntimeError: request license token faild (3)

**Comment by jondy** [comment, marked as answer]
You're using old license in Pyarmor 8.0, please check documentation to upgrade it.

**Reply by sana888999** [reply]
I just bought it for on 30-APR-2023  Pyarmor-8... it doesnt work with old version either i used web-ui to register but it didn't work... do i have to pay again to upgrade to use with Pyarmor-8?

**Reply by jondy** [reply]
Please check http://pyarmor.readthedocs.io/en/latest/licenses.html

**Comment by alikaz3mi** [comment]
After some files are obfuscated, I am facing this issue too.



```
Pyarmor 8.3.1 (group), 006070, dashboard

License Type    : pyarmor-group
License No.     : pyarmor-vax-006070
License To      : HEMSCap
License Product : dashboard

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Offline obfuscation
```



```
INFO     Python 3.8.10
INFO     Pyarmor 8.3.1 (group), 006070, dashboard
INFO     Platform linux.x86_64
INFO     search inputs ...
INFO     find package at dashboard_face_recognition
INFO     find 1 top resources
INFO     start to generate runtime files
INFO     target platforms {'linux.x86_64'}
INFO     write encoded_face_rec/pyarmor_runtime_006070/pyarmor_runtime.so
INFO     patch runtime file
INFO     generate runtime files OK
INFO     start to obfuscate scripts
INFO     process resource "dashboard_face_recognition"
INFO     obfuscating file config.py
INFO     write encoded_face_rec/dashboard_face_recognition/config.py
INFO     obfuscating file dashboard.py
INFO     write encoded_face_rec/dashboard_face_recognition/dashboard.py
INFO     obfuscating file __init__.py
...
...
INFO     obfuscating file logger.py
ERROR    request license token faild (3)
```

**Reply by jondy** [reply]
@alikaz3mi. It's strange, the group license should not request license token

How about use debug option `pyarmor -d gen ...` and check the log?

**Reply by jondy** [reply]
@alikaz3mi  But if hardware information of this machine is changed, group license may failed and will request license token.

**Reply by alikaz3mi** [reply]
I have tested in two devices. First I registered a new device file to them. pyarmor -v shows the correct information.

Both of the systems gave me the following message. Worth to say that one of them is my virtual machine.


```
INFO     obfuscating file contextmanagers.py
DEBUG    process inline marker
DEBUG    parse script
DEBUG    process co
DEBUG    patch co
DEBUG    unknown error, please check pyarmor.error.log
ERROR    request license token faild (3)
dashvm@dashvm-Standard-PC-i440FX-PIIX-1996:~/obfuscation$ cat pyarmor.error.log 
2715812 MainProcess 2023-08-13 09:08:32,720
Traceback (most recent call last):
  File "/home/dashvm/.local/lib/python3.8/site-packages/pyarmor/cli/__main__.py", line 702, in main
  File "/home/dashvm/.local/lib/python3.8/site-packages/pyarmor/cli/__main__.py", line 690, in main_entry
  File "/home/dashvm/.local/lib/python3.8/site-packages/pyarmor/cli/__main__.py", line 229, in cmd_gen
  File "/home/dashvm/.local/lib/python3.8/site-packages/pyarmor/cli/generate.py", line 156, in process
  File "/home/dashvm/.local/lib/python3.8/site-packages/pyarmor/cli/generate.py", line 119, in _obfuscate_scripts
  File "/home/dashvm/.local/lib/python3.8/site-packages/pyarmor/cli/core/__init__.py", line 103, in generate_obfuscated_script
  File "<maker>", line 665, in generate_obfuscated_script
  File "/home/dashvm/.local/lib/python3.8/site-packages/pyarmor/cli/__init__.py", line 16, in process
  File "<maker>", line 504, in process
  File "<maker>", line 510, in coserialize
  File "<maker>", line 555, in _build_ast_body
RuntimeError: request license token faild (3)****
```

**Reply by alikaz3mi** [reply]
Also, I tried the procedure without internet connection. It showed me this. 


```
INFO     Python 3.9.13
INFO     Pyarmor 8.3.1 (group), 006070, dashboard
INFO     Platform linux.x86_64
DEBUG    native platform linux.x86_64
DEBUG    home path: /home/deepdev/.pyarmor
DEBUG    command options: {'recursive': True, 'output': 'xxx', 'no_runtime': False, 'inputs': ['dashboard_face_recognition']}
DEBUG    install plugin: CodesignPlugin
DEBUG    install plugin: DarwinUniversalPlugin
INFO     search inputs ...
INFO     find package at dashboard_face_recognition
INFO     find 1 top resources
DEBUG    unknown error, please check pyarmor.error.log
ERROR    request license token faild (101)
(py39) deepdev@deepdev-H270-Gaming-3:~/PycharmProjects/FaceRecognition$ cat pyarmor.error.log 
196591 MainProcess 2023-08-13 09:14:58,914
Traceback (most recent call last):
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/__main__.py", line 702, in main
    main_entry(sys.argv[1:])
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/__main__.py", line 690, in main_entry
    return args.func(ctx, args)
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/__main__.py", line 229, in cmd_gen
    builder.process(options)
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/generate.py", line 147, in process
    Pytransform3.pre_build(self.ctx)
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/core/__init__.py", line 117, in pre_build
    m = Pytransform3.init(ctx)
  File "/home/deepdev/anaconda3/envs/py39/lib/python3.9/site-packages/pyarmor/cli/core/__init__.py", line 97, in init
    m.init_ctx(ctx)
RuntimeError: request license token faild (101)
```


Moreover, when I disable internet connection, my license information changes to trial mode and I have to re-register again. 


```
(py39) deepdev@deepdev-H270-Gaming-3:~/PycharmProjects/FaceRecognition$  pyarmor -v
Pyarmor 8.3.1 (trial), 000000, non-profits

License Type    : pyarmor-trial
License No.     : pyarmor-vax-000000
License To      : 
License Product : non-profits

BCC Mode        : No
RFT Mode        : No

Notes
* Can't obfuscate big script and mix str

(py39) deepdev@deepdev-H270-Gaming-3:~/PycharmProjects/FaceRecognition$ cd
(py39) deepdev@deepdev-H270-Gaming-3:~$ pyarmor reg pyarmor-device-regfile-6070.4.zip 
INFO     Python 3.9.13
INFO     Pyarmor 8.3.1 (trial), 000000, non-profits
INFO     Platform linux.x86_64
INFO     register "pyarmor-device-regfile-6070.4.zip"
INFO     This license registration information:

License Type    : pyarmor-group
License No.     : pyarmor-vax-006070
License To      : HEMSCap
License Product : dashboard

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Offline obfuscation
```

**Comment by jasonwang178** [comment]
Same issue in 8.3.1. I consistently used the same Ubuntu 22.04 server without any hardware alterations.

I'm using a group offline license:

```
Pyarmor 8.3.1 (group), 006072, Krugle

License Type    : pyarmor-group
License No.     : pyarmor-vax-006072
License To      : Aragon
License Product : Krugle

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Offline obfuscation
```

Command I used:

```shell
pyarmor cfg enable_trace=1
pyarmor cfg outer_keyname=trial.lic
pyarmor cfg package_name_format "core_runtime"
pyarmor gen --outer --enable-jit --mix-str -O dist2 -r dist/app
```

In fact, 50% of my files have been successfully obfuscated. But the rest stopped at the`request license token faild (3)` error.

![image](https://github.com/dashingsoft/pyarmor/assets/222802/e9edbedd-1e17-4f49-85dd-692dd607a74f)

The debugging logs:

```
2819123 MainProcess 2023-08-13 02:58:13,622
Traceback (most recent call last):
  File "/root/.pyenv/versions/3.10.12/lib/python3.10/site-packages/pyarmor/cli/__main__.py", line 702, in main
  File "/root/.pyenv/versions/3.10.12/lib/python3.10/site-packages/pyarmor/cli/__main__.py", line 690, in main_entry
  File "/root/.pyenv/versions/3.10.12/lib/python3.10/site-packages/pyarmor/cli/__main__.py", line 229, in cmd_gen
  File "/root/.pyenv/versions/3.10.12/lib/python3.10/site-packages/pyarmor/cli/generate.py", line 156, in process
  File "/root/.pyenv/versions/3.10.12/lib/python3.10/site-packages/pyarmor/cli/generate.py", line 119, in _obfuscate_scripts
  File "/root/.pyenv/versions/3.10.12/lib/python3.10/site-packages/pyarmor/cli/core/__init__.py", line 103, in generate_obfuscated_script
  File "<maker>", line 665, in generate_obfuscated_script
  File "/root/.pyenv/versions/3.10.12/lib/python3.10/site-packages/pyarmor/cli/__init__.py", line 16, in process
  File "<maker>", line 495, in process
  File "<maker>", line 495, in <listcomp>
  File "/root/.pyenv/versions/3.10.12/lib/python3.10/site-packages/pyarmor/cli/__init__.py", line 16, in process
  File "<maker>", line 881, in process
  File "<maker>", line 855, in mix_node
RuntimeError: request license token faild (3)
```

Additional tests I conducted:

* The functionality is successful upon omitting the --enable-jit --mix-str parameters.
* Executing `pyarmor gen --outer --enable-jit --mix-str INDIVIDUAL_FILE` for individual files produces the desired outcome.
* Uninstalled the pyarmor and removed `~/.pyarmor`, re-generated a new offline device reg file, and re-register it again on this device. Same error.
* Tried downgrade pyarmor to some prior version, i.e. 8.2.x, 8.3.0, but failed at installing the license. Seems the license file is incompatible?


I'm puzzled about why an offline group license still requires requesting the license token. How can I solve this issue? Thanks.

**Comment by jondy** [comment]
If group license machine id is not matched current device, it will request license token.

The common solution:
1. Upgrade Pyarmor to 8.3.0+
2. Check machine id in offline reg file `pyarmor-device-regfile-xxxx.1.zip`, unzip this file to any path `xxx`, and `ls xxx/tokens`
3. Check machine id of current machine
```
python
>>> from pyarmor.cli.core import Pytransform3 
>>> print('machine id is "%s"' % Pytransform3.get_hd_info(16))
```  
Make sure current device machine id is same as machine id in the offline reg file `pyarmor-device-regfile-xxxx.1.zip`

**Reply by jasonwang178** [reply]
The device ID discrepancy is intriguing. I adhered to the guide while creating the device registration file. However, I'm uncertain about the misstep in my process. Here's what I did:

1. On device B, I generated the group device file using the command `pyarmor reg -g 4`, resulting in `.pyarmor/group/pyarmor-group-device.4`.
2. Subsequently, I transferred the `.pyarmor/group/pyarmor-group-device.4` to my local macOS machine (device A), where I initially applied my pyarmor license using `pyarmor-regcode-6072.txt`.
3. On device A, I generated the offline device registration zip file using the command `pyarmor reg -g 4 pyarmor-regfile-6072.zip`.
4. Lastly, I moved the generated zip file back to device B and registered the `pyarmor-device-regfile-6072.4.zip` file there.

Although the tokens for `.pyarmor/group/pyarmor-group-device.4` and `pyarmor-device-regfile-6072.4.zip` match, they differ from the token produced when I executed the command on device B:

```py
Python 3.10.12 (main, Aug  8 2023, 09:15:11) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyarmor.cli.core import Pytransform3 
>>> print('machine id is "%s"' % Pytransform3.get_hd_info(13))
machine id is "b'dd39e721c41857fe78e1c4e0a9c4a750e'"
```

![image](https://github.com/dashingsoft/pyarmor/assets/222802/ac8e339e-11ca-48a8-952a-4d8c193dc57e)


The device B functions as a **Linux server hosted on AWS**. I've installed two Python versions using **pyenv**, which are as follows. And I'm using the version of 3.10.12.

```shell
pyenv versions
  system
* 3.10.12 (set by /home/ubuntu/.python-version)
  3.11.4
```

Both 3.10.12 and 3.11.4 have been installed with pyarmor, and since they share the same license, running `pyarmor -v` displays both versions as part of the offline group license.

```
Pyarmor 8.3.1 (group), 006072, Krugle

License Type    : pyarmor-group
License No.     : pyarmor-vax-006072
License To      : Aragon
License Product : Krugle

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Offline obfuscation
```

**Reply by jasonwang178** [reply]
I attempted the following steps, but they were unsuccessful:

I have generated a new `.pyarmor/group/pyarmor-group-device.6` file from device B. Subsequently, I manually adjusted the machine ID within the file, changing it from `gd39e721c41857fe78e1c4e0a9c4a750e` to `dd39e721c41857fe78e1c4e0a9c4a750e` (the ID obtained from `Pytransform3.get_hd_info(13)`). Afterward, I registered it from device A and attempted to install the device license 6. However, I encountered the following error message:

```
INFO     Python 3.10.12
INFO     Pyarmor 8.3.1 (group), 006072, Krugle
INFO     Platform linux.x86_64
INFO     register "pyarmor-device-regfile-6072.6.zip"
ERROR    could not get docker host machine id
Traceback (most recent call last):
  File "/root/.pyenv/versions/3.10.12/lib/python3.10/site-packages/pyarmor/cli/register.py", line 187, in _get_docker_hostname
    s.connect((host, port))
socket.gaierror: [Errno -2] Name or service not known
ERROR    could not get docker host machine id
```

It's important to note that this machine is a standard AWS EC2 instance and not a Docker container. Although Docker is installed. The Pyarmor actions are being executed from this machine (the Docker host) and not within any Docker containers. I'm unsure why the error message references not being able to obtain the Docker host machine ID.

**Reply by jasonwang178** [reply]
I discovered that executing `Pytransform3.get_hd_info(16)` enables me to obtain the correct ID for generating the `.pyarmor/group/pyarmor-group-device.6` file. However, this is not the case with `Pytransform3.get_hd_info(13)`.

![image](https://github.com/dashingsoft/pyarmor/assets/222802/4f8906be-6b92-4d08-9644-cd87904bfc93)

@jondy do you think this is a pyarmor bug? Thanks.

**Reply by jondy** [reply]
@jasonwang178 it should be `get_hd_info(16)`, `get_hd_info(13)` should not work.

**Reply by jasonwang178** [reply]
Hi @jondy, should I await the fix until the next release? Alternatively, is there a current workaround to address the issue?

**Reply by jondy** [reply]
8.3.1 should work, please first make sure `get_hd_info(16)` could get same machine id.

**Reply by jasonwang178** [reply]
In my case, the ID retrieved from `Pytransform3.get_hd_info(16)` matches the ID of my offline group device license, as evident from my previous comments.

**Reply by jondy** [reply]
@jasonwang178 If machine id is match, please report a bug in issues.

Please describe reproduced steps and necessary logs, do not paste snapshot.

The group license works in my test environments.

**Comment by alikaz3mi** [comment]
Dear @jondy
As you requested, I have reported the bugs I have faced during encoding process. 
I am facing the same issues as @jasonwang178. 

On the virtual machine device where I have generated device file id 3, the pyarmor-device-regfile-xxxx.3.zip token is equal to  ` Pytransform3.get_hd_info(16)`. Nevertheless the encoding procedure still doesn't work for large packages. And it doesn't work in offline mode. Although `pyarmor -v` shows all of the information about group license. The value is `gf4f51d4e1e8fb92c46b80fb891376c73`

On the other system where I initially used the registration file and generated offline device file 1. 
The value of pyarmor-device-regfile-6070.1.zip/token is `g9d3545b102003843f051b8b4db6e508c`.
While Python illustrates b'g063dec1216b8fda7779c8f3e32644ba8'. 

pyarmor-device-regfile-6070.2.zip/token value is ge21ceefcc843bb5f43a7a75129cecf49. The value for device id 4 is g063dec1216b8fda7779c8f3e32644ba8. Which is equal to python value. 

Device ids 1, 2, and 4 are on the same device. Device id 3 is on another device.

**Comment by jondy** [comment]
If using group license in virtual machine, computer name, network and hard disk can't be changed. Reboot virtual machine, check machine id and make sure they're same

```
python
>>> from pyarmor.cli.core import Pytransform3 
>>> print('machine id is "%s"' % Pytransform3.get_hd_info(16))
```  

Do not use `Pytransform3.get_hd_info(13)`, it's mistake.

**Comment by jondy** [comment]
@jasonwang178 

Could you first apply this patch and check the log?

```patch
--- a/pyarmor/cli/register.py
+++ b/pyarmor/cli/register.py
@@ -153,8 +153,11 @@ class Register(object):
                 f.extract(item, path=path)
             namelist = f.namelist()
             if 'group.tokens' in namelist:
+                logger.info('group license machine id: %s',
+                             [x for x in namelist if x.startswith('tokens')])
                 for idver in (16, 11):
                     machid = self._get_machine_id(idver).decode('utf-8')
+                    logger.info('got machine id: %s', machid)
                     name = '/'.join(['tokens', machid])
                     if name in namelist:
                         break
```

**Comment by jondy** [comment]
I'll lock this thread because it's out of original problem. 

Current problem is group license problem, it's different from the original problem, and too many threads make it too confusion

**Comment by jondy** [comment]
There is really a bug for running unlimited docker, apply this patch to fix it

```patch
--- a/pyarmor/cli/register.py
+++ b/pyarmor/cli/register.py
@@ -190,7 +190,7 @@ class Register(object):
                 s.connect((host, port))
                 s.sendall(b'PADH' + b'x' * 60)
                 flag = s.recv(1)
-                if flag in (b'a', b'b'):
+                if flag in (b'a', b'b', b'g'):
                     data = s.recv(32)
                 machid = (flag + data).decode('utf-8')
                 logger.info('got docker host machine id: %s', machid)
```

**Comment by jondy** [comment]
@jasonwang178 If the issue is related to run unlimited docker, there is really a bug. I'll publish a new release 8.3.2 in 2 days to fix this issue.
